src/main/java/org/broadinstitute/hellbender/utils/runtime/StreamingProcessController.java

```diff
@@ -0,0 +1,598 @@
+                    }
+                } else {
+                    // We didn't timeout on stderr, try to get stdout. If its not immediately ready, use a timeout
+                    // and wait for it.
+                    if (stdOutFuture != null) {
+                        if (stderr == null) {
+                            // no stderr timeout; no stderr, and stdout isn't done, so use our timeout on stdout
+                            stdout = stdOutFuture.get(timeOutMillis, TimeUnit.MILLISECONDS);
+                        } else {
+                            // stderr didn't timeout, stdout isn't ready, but we already have stderr output, so fall
+                            // through and just return that it
+                        }
+                    } else {
+                        throw new TimeoutException("No stdout or stderr was available. The timeout period was exceeded.");
+                    }
+                }
+            } catch (TimeoutException e) {
+                // we didn't get any stderr, but thats ok since we got stdout
+            } catch (InterruptedException e) {
+                throw new GATKException("InterruptedException retrieving stderr", e);
+            } catch (ExecutionException e) {
+                throw new GATKException("ExecutionException retrieving stderr", e);
+            }
+        }
+
+        //log the output, and restart the listeners for next time
+        processJournal.writeInbound(stdout, stderr);
+        startListeners();
+
+        return new ProcessOutput(0, stdout, stderr);
+    }
+
+    /**
+     * Create a temporary FIFO suitable for sending output to the remote process. The FIFO is only valid for the
+     * lifetime of the controller; the FIFO is destroyed when the controller is destroyed.
+     *
+     * @return a FIFO File
+     */
+    public File createFIFO() {
+        if (fifoTempDir != null || fifoFile != null) {
+            throw new IllegalArgumentException("Only one FIFO per controller is supported");
+        }
+
+        fifoTempDir = Files.createTempDir();
+        final String fifoTempFileName = String.format("%s/%s", fifoTempDir.getAbsolutePath(), "gatkStreamingController.fifo");
+
+        // create the FIFO by executing mkfifo via another ProcessController
+        final ProcessSettings mkFIFOSettings = new ProcessSettings(new String[]{"mkfifo", fifoTempFileName});
+        mkFIFOSettings.getStdoutSettings().setBufferSize(-1);
+        mkFIFOSettings.setRedirectErrorStream(true);
+        final ProcessController mkFIFOController = new ProcessController();
+        final ProcessOutput result = mkFIFOController.exec(mkFIFOSettings);
+        final int exitValue = result.getExitValue();
+
+        fifoFile = new File(fifoTempFileName);
+        if (exitValue != 0) {
+            throw new GATKException(String.format("Failure creating FIFO (%s) got exit code (%d)", fifoTempFileName, exitValue));
```

Comment:
If this triggers it would be really helpful to dump the stderr/ stdout from the process as part of the error.

Reply:
Done.